confmap/confmap.go

```diff
@@ -431,6 +432,40 @@ func unmarshalerEmbeddedStructsHookFunc() mapstructure.DecodeHookFuncValue {
 	}
 }
 
+// optionalHookFunc applies logic when the to.Type is optional.Optional. When decoding primitive types, 
+// we use reflect.Value's FieldByName API (https://pkg.go.dev/reflect#Value.FieldByName) and Set (https://pkg.go.dev/reflect#Value.Set) APIs
+// in order to set the Value and HasValue. Set cannot be used for struct field types, as the set value must be assignable to the target type, 
+// so for structs we use optionals custom unmarshaller in order to cast the map to the struct.
+//
+// This logic relies on the fact that the hook only gets called when the value is explicitely set in the config,
+// so if we are in this hook, we know we can set HasValue to true.
+func optionalHookFunc() mapstructure.DecodeHookFuncValue {
+	return func(from reflect.Value, to reflect.Value) (any, error) {
+		// As the type is generic, we check for the prefix.
+		if strings.HasPrefix(to.Type().String(), "optional.Optional") {
+			// the optional.Optional field is a struct
+			if _, ok := from.Interface().(map[string]any); ok {
+				unmarshaler, ok := to.Addr().Interface().(Unmarshaler)
+				if !ok {
+					return from.Interface(), nil
+				}
+				c := NewFromStringMap(from.Interface().(map[string]any))
+				if err := unmarshaler.Unmarshal(c); err != nil {
+					return nil, err
+				}
+				return to.Interface(), nil
+			}
+
+			// the optional.Optional field is a primitive type
```

Comment:
I'm just spitballing here, but would it be possible to instead have an interface specifically for unmarshalling primitive values? For example, maybe someone would want to do the following:

```golang
type EsotericIntEncoding int

func (e *EsotericIntEncoding) UnmarshalPrimitive(val any) error {
	x, ok := val.(int)
	if !ok {
		return errors.New("not an int")
	}
	*e = EsotericIntEncoding(x)
	return nil
}
```

We could then check if `from` is not a map and if `to` implements the `PrimitiveUnmarshaler` interface and unmarshal things like `Optional[int]` this way. The interface would also let component authors unmarshal custom types that alias primitive values like `EsotericIntEncoding`.

Reply:
This would also allow the `Optional` type to handle this value itself and to avoid needing to make its fields public so they can be accessed by reflection.

Reply:
> I'm just spitballing here, but would it be possible to instead have an interface specifically for unmarshalling primitive values

I'm not sure to understand how this would work:
- Seems like the receiver would need to be `Optional[T]` if we want to set the `hasValue` field ?
- With this approach, would we need to have a separate `UnmarshalPrimitive` func for each type we want to support ?

> This would also allow the Optional type to handle this value itself and to avoid needing to make its fields public so they can be accessed by reflection.

I've made changes that unexport the Optional structs fields, by using a custom unmarshaller.

Reply:
>Seems like the receiver would need to be Optional[T] if we want to set the hasValue field ?

Yeah, I think an `*Optional[T]` receiver is the way to go. The way you currently have it looks good to me.

>With this approach, would we need to have a separate UnmarshalPrimitive func for each type we want to support ?

Yes, I think it could help with other primitive types (e.g. special encodings for strings, ints, etc.) in addition to helping here with `Optional[T]` where `T` is a primitive type. It also means we don't need a special hook for only our type if we put logic like the following in `unmarshalerHookFunk`:

```golang
if _, ok = from.Interface().(map[string]any); !ok {
	unmarshaler, ok := toPtr.(PrimitiveUnmarshaler)
	if !ok {
		return from.Interface(), nil
	}
	if err := unmarshaler.UnmarshalPrimitive(from.Interface()); err != nil {
		return nil, err
	}
	return unmarshaler, nil
}
```

I tested this locally and it appears to work.

Reply:
I would like input from both of them before continuing. We should also check whether @yurishkuro would like to continue work on #10260, since that was opened before this.

If you have time, would you be willing to add/update some basic tests, or report on the output of manually running the Collector with these changes, to show that this is viable?

Reply:
I will mention as an aside that I looked at `encoding.TextUnmarshaler` as an alternative that wouldn't require us to provide a new interface, but ultimately we want to use the string -> primitive type unmarshaling provided by mapstructure, so in the end I think this would add more complexity than it removes. Implementing our own interface allows us to accept partially-unmarshaled values and do any final transformations on them.

The downside here is that `encoding.TextUnmarshaler` and our `PrimitiveUnmarshaler` interfaces would have some overlap, and users would need to know which one to use. I also don't have any immediate use-cases for `PrimitiveUnmarshaler` in mind outside this optional type.

Reply:
> If you have time, would you be willing to add/update some basic tests, or report on the output of manually running the Collector with these changes, to show that this is viable?

I added a `collector.yaml` and [print statements](https://github.com/open-telemetry/opentelemetry-collector/pull/11409/files#diff-1fadfcd3b4194004717150126d27e8a32fdaa8292ca99e81a937d26abe0998faR153) in this PR which demonstrates the results, see output:
```
2024-10-16T14:48:51.880+0200	info	service/service.go:135	Setting up own telemetry...
2024-10-16T14:48:51.880+0200	info	telemetry/metrics.go:70	Serving metrics	{"address": "localhost:8888", "metrics level": "Normal"}
2024-10-16T14:48:51.881+0200	info	service/service.go:207	Starting otelcorecol...	{"Version": "0.111.0-dev", "NumCPU": 10}
2024-10-16T14:48:51.881+0200	info	extensions/extensions.go:39	Starting extensions...
MaxIdleConns: {value:567 hasValue:true}
MaxIdleConnsPerHost: {value:0 hasValue:false}
AUTH: {value:{Authentication:{AuthenticatorID:{typeVal:{name:} nameVal:}} RequestParameters:[]} hasValue:false}
CORS: {value:{AllowedOrigins:[http://test.com] AllowedHeaders:[Example-Header] MaxAge:7200} hasValue:true}
2024-10-16T14:48:51.881+0200	info	otlpreceiver/otlp.go:169	Starting HTTP server	{"kind": "receiver", "name": "otlp", "data_type": "traces", "endpoint": "localhost:4317"}
2024-10-16T14:48:51.881+0200	info	service/service.go:230	Everything is ready. Begin running and processing data.
```

let me know if this is enough, or if I should still add tests ?

Reply:
> We should also check whether @yurishkuro would like to continue work on https://github.com/open-telemetry/opentelemetry-collector/pull/10260, since that was opened before this.

Happy to close this in favor of Yuri's PR. The only additions/substraction I made here:

- Remove `defaultFn` field in Optional struct, I don't think this is required
- Add another unmarshaller. 10260 errors for optional types that contain a primitive value, as they cannot be cast to a map. (see this comment: https://github.com/open-telemetry/opentelemetry-collector/pull/10260#issuecomment-2371516921)